content/plugin/log/writing.mdx

```diff
@@ -0,0 +1,179 @@
+
+!> **Warning:** Do not use `fmt.Println` and similar methods to log. They will silently fail, making the output unavailable.
+
+
+## Structured Logging
+
+The `tflog` package uses structured logging, based on [`go-hclog`](https://pkg.go.dev/github.com/hashicorp/go-hclog). Rather than writing logs as sentences with embedded variables and values, `tflog` takes a sentence describing the logging event and a set of variables to log. When variables are separate from the log description, you can use them to programmatically parse, filter, and search log output. This separation also allows other parts of the system to associate variables with downstream log output.
+
+## Log Levels
+
+You must choose a verbosity level for each line of log output. This lets consumers [specify a type of log output](/plugin/log/managing#log-levels) to write from your provider. For example, you can use environment variables to set your provider to write only logs of type `Warn` during a Terraform run.
+
+### Error
+
+The least verbose output that typically describes an unexpected condition prior to halting execution. It often provides more information about a user-facing error.
+
+```go
+tflog.Error(ctx, "Unrecognized API response body")
+```
+
+### Warn
+
+Output that describes an unexpected condition, but not one that should halt execution. It often includes deprecations or environment issues.
+
+```go
+tflog.Warn(ctx, "Retrying due to API server-side error")
+```
+
+### Info
+
+Output that describes a certain logic condition or event. It often includes details about the environment your provider is running in or how it has been configured to run.
+
+```go
+tflog.Info(ctx, "Using API token for authentication")
+```
+
+### Debug
+
+Verbose output that typically describes important operational details like milestones in logic. It often describes behaviors that may confusing even though they are correct.
+
+```go
+tflog.Debug(ctx, "Two identical diagnostics in the response, deduplicating down to one")
+```
+
+### Trace
+
+The most verbose output that describes the lowest level operational details, such as intra-function steps or raw data.
+
+```go
+tflog.Trace(ctx, "Creating the widget")
+```
+
+## Variables
+
+Use variables to attach information to specific logs or to an entire logger, which adds that information to all subsequent logs. You can also combine both methods to simplify your logging code.
+
+### Single Log Variables
+
+To specify filterable variables in the log output, add sets of key and value argument pairs after the log message. You can add as many key and value argument pairs as you need, but the key must always come before its associated value.
+
+The following example adds both a URL and a method for an API request.
+
+```go
+tflog.Trace(ctx, "executing API request",
+    "url", "https://www.example.com/my/endpoint", "method", "POST")
```

Comment:
Nit: Might be nice to newline split these based on pairings for clarity 😄 

```suggestion
    "url", "https://www.example.com/my/endpoint",
    "method", "POST")
```

Reply:
yes good point!